Fixed not reading local config under ~/.aws/credentials

sashokbg opened the pull request:
Fixed a problem #52  where the app will not connect to the appropriate local dynamodb instance when local credential files are configured in ~/.aws/credentials

rchl commented:
1. The idea is to avoid modifying global environment with `AWS.config.update()` and avoid polluting environment variables (with `AWS_SDK_LOAD_CONFIG`) to configure AWS package because this project can be used programatically as a library and it's a very bad idea to modify environment like that. You can always avoid those by creating instances of credential objects and passing them to `AWS.DynamoDB` constructor like was the case before.

2. Instead of setting `AWS_SDK_LOAD_CONFIG` env variable, you can just create credentials from file using `new AWS.SharedIniFileCredentials({profile: 'myprofile'});`.

3. You can see above that `AWS.SharedIniFileCredentials` constructor accepts `profile` option. This is important to have in mind because it's not always ideal to load `default` profile. It might not even exist as in my case.
Here we have to think a bit more how to handle that cleanly. It's a bit hard to grasp given how many ways there are to set credentials.

4. Please follow project's style. No semicolons and 2 spaces for indentation. (Not a fan myself of no semicolons but I'm not the original author of the project).

rchl commented:
5. I believe this change would make it impossible to override KEY/SECRET with ones provided on the command line if one has aws configuration file.

sashokbg commented:
Hello @rchl 
Thank you for your feedback on my PR. If I understood correctly, the main concern here is that we should avoid updating the environment variables of the AWS sdk.

I took your comments into account and created a second commit to improve the PR.

I am now only reading the AWS.config and AWS.config.credentials variables and instead of updating the env, I create a new object that is then passed to the dynamodb client.

The order of variable loading is as follows:
1) We check if local config is available in AWS.config / AWS.config.credentials. If true - use it
2) We check if env variables env.AWS_REGION and env.AWS_ACCESS_KEY_ID are present. If true - override 1)
3) If none of the above - use defaults ('us-east-1' and 'key')

I also completely removed the client secret since the documentation states that only the region and the key are used to determine the dynamodb instance - more on that [here](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.UsageNotes.html)

I added a small log that show which region / client key are used when starting up the application to avoid confusion

rchl commented on lib/backend.js:
```diff
@@ -9,16 +9,52 @@ const pickBy = require('lodash/pickBy')
-
 require('es7-object-polyfill')
 
+process.env.AWS_SDK_LOAD_CONFIG=1
```
See my second point from previous comment. I'd like to avoid modifying global environment and there is an API to read credentials from file without setting this variable.

rchl replied:
Let me have a look again.

sashokbg replied:
I stumbled upon a doc which explains how the AWS_SDK_LOAD_CONFIG var works.
https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/setting-region.html

It looks to me that we can set this var without any concern as it is the last in the order of precedence for loading the region.

So even if we set the AWS_SDK_LOAD_CONFIG to true this will not effect an existing project since they will most likely override AWS_REGION using an env variable or programmatically.

What are your thoughts on this ?

sashokbg replied:
I checked the SDK and it seems to load automatically the credentials file but not the config file (region is in ~/.aws/config and not in ~/.aws/credentials). I even went on looking at the sources. 

You have a valid point about the lib use case, but I think if we use it as a lib we would already set the region programmatically or via env var. Or worse if no region is set we end up with default region us-east-1. So nothing should be broken but we at least fix the problem for people who wanted to use it as a standalone tool (most of the use cases ?)

rchl replied:
OK, then we have no option but to keep it.

But please move this env assignment and require('aws-sdk') back to where it was initially so that we don't unnecessarily trigger loading when using this programatically.

rchl commented on lib/backend.js:
```diff
@@ -9,16 +9,52 @@ const pickBy = require('lodash/pickBy')
-    accessKeyId: env.AWS_ACCESS_KEY_ID || 'key', // DynamoDB Local doesn't care what the key/secret are
-    secretAccessKey: env.AWS_SECRET_ACCESS_KEY || 'secret'
-  }
+  const dynamoConfig = loadDynamoConfig(env)
```
Could you move config reading after the if/else block below? There is no point in trying to read config if we fail the amazonaws.com check below.

rchl commented on lib/backend.js:
```diff
@@ -9,16 +9,52 @@ const pickBy = require('lodash/pickBy')
+    endpoint: 'http://localhost:8000',
+    sslEnabled: false,
+    region: 'us-east-1',
+    accessKeyId: 'key'
```
So you left out the secret. I understand that this is based on reading documentation but in practice it makes it impossible to access DB in my case.

I get this error when trying to load index page:
```
{
  "error": {
    "message": "Missing credentials in config",
    "code": "CredentialsError",
    "time": "2018-10-01T07:48:43.576Z",
    "retryable": true,
    "originalError": {
      "message": "Could not load credentials from any providers",
      "code": "CredentialsError",
      "time": "2018-10-01T07:48:43.576Z",
      "retryable": true,
      "originalError": {
        "message": "Connection timed out after 1000ms",
        "code": "TimeoutError",
        "time": "2018-10-01T07:48:43.575Z",
        "retryable": true
      }
    }
  }
}
```

My case is:
 - I have `~/.aws/credentials` file but it doesn't include `default` profile credentials so it's not really used to access my local DB
 - My local DB is started with `/usr/local/bin/dynamodb_control.py --inmemory no http://localhost:8000`

It works if I add the `secretAccessKey` key to the dynamo config object. The value of the key doesn't matter as long as it's non-empty.

sashokbg replied:
Ok so in either case we need a secret key set to some constant value. I will fix this in a sec

sashokbg commented:
Hello, sorry for the delay on my end I had a hard week at work.
I moved the loadDynamoConfig after the amazonaws.com check and I added a default secretAccessKey.

Could you please help me do a checklist of what needs to be done from no on, because I am getting a bit lost with all the comments we had.

Thank you :)

rchl commented:
- [ ] Move `process.env.AWS_SDK_LOAD_CONFIG=1` and `const AWS = require('aws-sdk')` to after the `if (!dynamodb || !docClient) {` check so that we don't modify environment and load AWS when not necessary. As in, instead of causing side effects when library is required, only do that when code path requires it. You can pass AWS object through `createAwsConfig()` function.
- [ ] Maybe re-add possibility to override `secret` with environment variable. Don't know if that would be useful but doesn't hurt to have it.

sashokbg commented:
Hello,

- [x] Move process.env.AWS_SDK_LOAD_CONFIG=1 and const AWS = require('aws-sdk') to after the if (!dynamodb || !docClient)
I have moved the AWS init after the check that you mentioned.

- [x] Re-add possibility to override secret with environment variable
Done

- [x] I also did a small refactoring and extracted the code that checks the endpoint "env.DYNAMO_ENDPOINT.indexOf('.amazonaws.com') > -1" in its own function (preventing a potential bug)


I did a couple of tests locally and things seem to work.

rchl commented:
BTW. Will do a major version change as this can result in not being able to access database if someone has `default` credentials in ~/.aws/credentials but needs to use different ones for local database.

rchl commented:
With this fix included, I'm getting this error in my docker container running on anapsix/alpine-ava:8 base:

```

2018-10-15 11:15:25,360 DEBG 'dynamo-admin' stderr output:




6


  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);


                 ^




Error: ENOENT: no such file or directory, open '/root/.aws/credentials'


    at Object.fs.openSync (fs.js:646:18)


    at Object.fs.readFileSync (fs.js:551:33)


    at Object.readFileSync (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/util.js:97:26)


    at SharedIniFile.loadFile [as ensureFileLoaded] (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/shared_ini.js:19:18)


    at SharedIniFile.loadProfile [as getProfile] (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/shared_ini.js:52:10)


    at Config.region (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/node_loader.js:94:34)


    at Config.set (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/config.js:450:39)


    at Config.<anonymous> (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/config.js:285:12)


    at Config.each (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/util.js:485:32)


    at new Config (/usr/lib/node_modules/dynamodb-admin/node_modules/aws-sdk/lib/config.js:284:19)
```

rchl commented:
Fixed in 3.0.1 with 84010954c24fdac35f85e7f98224b437dae59c4e
